### ISSUE TITLE
[Android] PickFile() never returns, and subsequent calls trigger "Only one operation can be active at a time"

### ISSUE THREAD

**Issue by pingzing** [jfversluis/FilePicker-Plugin-for-Xamarin-and-Windows#28]
Hi! I'm trying to use version 2.0.8-beta of this library with Xamarin.Forms, version 2.5.0.122203. My cross-platform project is a .NET Standard 2.0 library.

Unfortunately, when I attempt to user the `PickFile()` method, it never returns anything:

```C#
private async void AddFile()
{
    var pickedFile = await _filePicker.PickFile();
    if (pickedFile == null) // This never runs!
    {
        return;
    }
}
```

In the above snippet, the `_filePicker` is injected via dependency injection, but ultimately comes from the `CrossFilePicker.Current` singleton.

Attempting to call this method a second time results in a `System.InvalidOperationException: Only one operation can be active at a time`.

Curiously, those are the results that I see on a Google Pixel running Android 8.1.0. In my GenyMotion emulator using an image similar to a Google Nexus 5, running Android 7.1.1, `PickFile()` DOES return, but the data `byte[]` array is always empty.

You can see a reproduction of this in my project at [its current commit, 2dd27d698e1d845e8eb6b1a53a25906e77903734](https://github.com/pingzing/Codeco/commit/2dd27d698e1d845e8eb6b1a53a25906e77903734). To reproduce the issue, simply boot the app, press the "ADD" button at the top right, and attempt to select the `test.csv` file found at the root of the repo (after moving it onto the phone or emulator, of course).

**Comment by jfversluis** [comment]
Thanks for the extensive report and reproduction! I will have a look as soon as I have some time available.

**Comment by pingzing** [comment]
Some more information: I cloned the repo, to do some debugging of my own. It looks like, ultimately, there were two issues:

1.) My app didn't have Storage permissions, despite declaring them in the manifest. I had to request them at runtime. This caused the plugin to throw an UnauthorizedAccessException when it tried to read the picked file.

2.) [On line 59 of FilePickerImplementation.cs](https://github.com/jfversluis/FilePicker-Plugin-for-Xamarin-and-Windows/blob/88f5e43cbf963a7a561f773d46497b226a222483/src/Plugin.FilePicker/Plugin.FilePicker.Android/FilePickerImplementation.cs#L59), any exception thrown in the lambda that gets passed to the `FileData` constructor won't get caught. I think that's what was causing my "never returns from the await" issue.

I've since added runtime permission requests for READ_EXTERNAL_STORAGE, and that fixed the issue.

A suggestion: would it be in scope for this plugin to do permission checking before it attempts to access the filesystem? Or is that something that should be left up to the application developer? Xamarin has some [decent documentation](https://developer.xamarin.com/guides/android/application_fundamentals/permissions/) on how to do it on every request.

Finally--I think there needs to be exception handling surrounding the lambda that the `FileData` constructor takes.

**Comment by vividos** [comment]
@jfversluis Starting from Android 6.0, adding permission READ_EXTERNAL_STORAGE to the AndroidManifest.xml isn't enough. At runtime the permission must be checked. Some permissions are "acknowledged" automatically by Android, whereas for some permissions, the user has to be asked before.

We could add requesting the permission automatically to the Android implementation. This would fix the "bugs" reported by users that forgot requesting the permissions, but takes away control from the user. I could also add a description to the README.md how permission checking could be done. What do you think we should do?

**Comment by vividos** [comment]
We just released NuGet package 2.0.106-beta with some Android related fixes. I also put in a check if the permission is actually granted. You have to request the permission from the user, though. See sample code:

https://github.com/jfversluis/FilePicker-Plugin-for-Xamarin-and-Windows/blob/develop/samples/Plugin.FilePicker.Sample.Forms/Forms/MainPage.xaml.cs

Be also sure to catch all exceptions from PickFile(), as it's the only way how the file picker can report errors to you. I also added an explanation for the "Only one operation can be active at a time" error on this project's README that should help.
